src/category_theory/limits/shapes/pullbacks.lean

```diff
@@ -0,0 +1,247 @@
+-- failure in `span`, below.
+lemma hom_id (X : walking_span.{v}) : hom.id X = 𝟙 X := rfl
+
+instance : small_category.{v+1} walking_span.{v} := sparse_category
+
+end walking_span
+
+open walking_span walking_cospan walking_span.hom walking_cospan.hom
+
+variables {C : Sort u} [𝒞 : category.{v+1} C]
+include 𝒞
+
+def cospan {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) : walking_cospan.{v} ⥤ C :=
+{ obj := λ x, match x with
+  | left := X
+  | right := Y
+  | one := Z
+  end,
+  map := λ x y h, match x, y, h with
+  | _, _, (id _) := 𝟙 _
+  | _, _, inl := f
+  | _, _, inr := g
+  end }
+def span {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) : walking_span.{v} ⥤ C :=
+{ obj := λ x, match x with
+  | zero := X
+  | left := Y
+  | right := Z
+  end,
+  map := λ x y h, match x, y, h with
+  | _, _, (id _) := 𝟙 _
+  | _, _, fst := f
+  | _, _, snd := g
+  end }
+
+@[simp] lemma cospan_left {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) :
+  (cospan f g).obj walking_cospan.left = X := rfl
+@[simp] lemma span_left {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) :
+  (span f g).obj walking_span.left = Y := rfl
+
+@[simp] lemma cospan_right {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) :
+  (cospan f g).obj walking_cospan.right = Y := rfl
+@[simp] lemma span_right {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) :
+  (span f g).obj walking_span.right = Z := rfl
+
+@[simp] lemma cospan_one {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) :
+  (cospan f g).obj walking_cospan.one = Z := rfl
+@[simp] lemma span_zero {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) :
+  (span f g).obj walking_span.zero = X := rfl
+
+@[simp] lemma cospan_map_inl {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) :
+  (cospan f g).map walking_cospan.hom.inl = f := rfl
+@[simp] lemma span_map_fst {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) :
+  (span f g).map walking_span.hom.fst = f := rfl
+
+@[simp] lemma cospan_map_inr {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) :
+  (cospan f g).map walking_cospan.hom.inr = g := rfl
+@[simp] lemma span_map_snd {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) :
+  (span f g).map walking_span.hom.snd = g := rfl
+
+@[simp] lemma cospan_map_id {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) (w : walking_cospan) :
+  (cospan f g).map (walking_cospan.hom.id w) = 𝟙 _ := rfl
+@[simp] lemma span_map_id {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) (w : walking_span) :
+  (span f g).map (walking_span.hom.id w) = 𝟙 _ := rfl
+
+
+variables {X Y Z : C}
+
+attribute [simp] walking_cospan.hom_id walking_span.hom_id
+
+def pullback_cone (f : X ⟶ Z) (g : Y ⟶ Z) := cone (cospan f g)
+
+namespace pullback_cone
+variables {f : X ⟶ Z} {g : Y ⟶ Z}
+
+def π₁ (t : pullback_cone f g) : t.X ⟶ X := t.π.app left
+def π₂ (t : pullback_cone f g) : t.X ⟶ Y := t.π.app right
+
+def mk {W : C} (π₁ : W ⟶ X) (π₂ : W ⟶ Y)
+  (eq : π₁ ≫ f = π₂ ≫ g) :
+  pullback_cone f g :=
+{ X := W,
+  π :=
+  { app := λ j, walking_cospan.cases_on j π₁ π₂ (π₁ ≫ f),
+    naturality' := λ j j' f, by cases f; obviously } }
+
+def condition (t : pullback_cone f g) : (π₁ t) ≫ f = (π₂ t) ≫ g :=
```

Comment:
make this a `lemma` (also `pushout_cocone.condition`)

Reply:
Fixed, thanks.